RFCs/symbol_alias.md

```diff
@@ -0,0 +1,171 @@
+* nim doesn't have a way to alias symbols, and the workaround using `template foo2(args: varargs[untyped]): untyped = foo(args)` falls short in many cases; eg
+  - if all parameters are optional, this won't work, see `case1`
+  - this requires different syntax for different symbols and doesn't work for some symbols, see `case2`
+  - it only provides an alternative way to call a routine but doesn't alias the symbol, so for example using introspection (eg to get parameter names) would give different results, see `case3`
+
+## Motivation 2: passing symbols to routines
+
+* nim doesn't allow passing templates or uninstantiated generics (or macros, iterators etc) to routines, so it forces you to either turn the callee into a template/macro, leading to much more complex, hard to debug code code (eg: see zero-functional https://github.com/zero-functional/zero-functional/blob/master/zero_functional.nim, with macros calling macros calling macros...).
+
+## Motivation 3: speed benefit
+Passing functors via closures is not free, because closure prevents inlining. Instead this RFC allows passing templates directly which is a zero-cost abstraction. eg: see `case4`.
+
+## Motivation 3: lambdas
+* this allows defining lambdas, which are more flexible than `sugar.=>`, and don't incur overhead, see `tests/magics/tlambda.nim`; eg: `case5`
+
+## Motivation 4: composable iterators
+composable iterators, which allows define lazy functional programming primitives (eager `toSeq` and lazy `map/filter/join` etc; which can have speed benefit compared to `sequtils`); the resulting code is quite nice and compact, see `tlambda_iterators.nim` which uses library `lambdaIter` that builds on top of `lambda`; it allows defining primitives that work regardless we're passing a value (`@[1,2,3]`) or a lazy iterate (eg `iota(3)`)
+
+## Motivation 5: this fixes or closes a lot of issues
+see a sample here: https://github.com/nim-lang/Nim/pull/11992
+
+## alias syntax: `alias foo2 = expr`
+
+```nim
+alias foo2 = expr # expr is an expression resolving to a symbol
+# eg:
+alias echo2 = echo # echo2 is the same symbol as `echo`
+echo2() # works
+echo2 1, "bar" # works
+
+alias echo2 = system.echo # works with fully qualified names
+
+import strutils
+alias toLowerAscii2 = strutils.toLowerAscii # works with overloaded symbols
+alias strutils2 = strutils # can alias modules2
+var z = 1
+alias z2 = z # works with var/let/const
+```
+
+## passing alias to a routine / generic parameter
+```nim
+proc fn(a: symbol) = discard # fn can be any routine (template etc)
+fn(alias echo) # pass symbol `echo` to `fn`
+proc fn(a, b: symbol) = discard # a, b are bind-many, not bind-once, unlike `seq`; there would be little use for bind-once
+```
+* a `symbol` parameter makes a routine implicitly generic
+* a `symbol` parameter matches a generic parameter:
+```nim
+proc fn[T](a: T) = discard
+fn(12) #ok
+fn(alias echo) #ok
+```
+
+## alias parameters are resolved early
+```nim
+proc fn(a: symbol) =
```

Comment:
Is this an implicit generic proc? If so, you need to write it out in this RFC.

Reply:
i did, see a few lines below:

> `* a `symbol` parameter makes a routine implicitly generic`